### PR TITLE
Fixed #29534 -- Made dbshell use rlwrap on Oracle if available.

### PR DESCRIPTION
**Pull request by danizen**
Design of the code is that another wrapper can easily be used and the way to wrap it can easily be modified.

### CODE REVIEW

**Comment by timgraham** [comment]
@felixxm, any interest in giving this some feedback or an initial review?

File: tests/dbshell/test_oracle.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import os
+import signal
+from unittest import mock
+
+from django.db.backends.oracle.client import DatabaseClient
+from django.test import SimpleTestCase
+
+
+class MockConnection(object):
```
Comment (felixxm) [inline tests/dbshell/test_oracle.py]: Do we need to mock connection? I think it is not necessary, you can use `django.db.connection`.

File: tests/dbshell/test_oracle.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            return '/usr/bin/rlwrap' if have_rlwrap else None
+
+        self.subprocess_args = None
+        with mock.patch('subprocess.call', new=_mock_subprocess_call):
+            with mock.patch('shutil.which', new=_mock_shutil_which):
+                client.runshell()
+        return self.subprocess_args
+
+    def test_basic(self):
+        connectstring = 'USER/PASSWORD@TNSNAME'
```
Comment (felixxm) [inline tests/dbshell/test_oracle.py]: IMO you can use `django.db.connection._connect_string()`.

File: docs/releases/2.1.txt [hunk 3]
```diff
@@ -126,6 +126,12 @@ CSRF
 * Added the :setting:`CSRF_COOKIE_SAMESITE` setting to set the ``SameSite``
   cookie flag on CSRF cookies.
 
+Database Backends
```
[the remark this answers is not in the document]
Comment (danizen) [reply docs/releases/2.1.txt]: I will do these things.

**Comment by felixxm** [comment]
@danizen Do you have time to finish this?

**Comment by felixxm** [comment]
Updated in #10555.

**Comment by aiswaryasankar** [comment]
**🔒 Entelligence AI Vulnerability Scanner**

✅ **No security vulnerabilities found!**

Your code passed our comprehensive security analysis.

📊 **Files Analyzed**: 1 files

---

**Comment by felixxm** [comment]
@aiswaryasankar Please stop leaving these pointless comment or I will be forced to block your account.

**Pull request closed without merging**